### DISCUSSION TITLE
Correctness of Grafana Dashboard Queries

### DISCUSSION THREAD

**Discussion by sthaha** [sustainable-computing-io/kepler#946] [General]
This discussion intents to validate the correctness of promqls used in Grafana Dashboard. 
A lot of promqls follows the pattern below
```promql
  sum(
    (increase(kepler_container_package_joules_total{container_namespace=~\"$namespace\"}[1h]) * $watt_per_second_to_kWh) 
    *
    (
      count_over_time(kepler_container_package_joules_total{container_namespace=~\"$namespace\"}[24h])
      /
      count_over_time(kepler_container_package_joules_total{container_namespace=~\"$namespace\"}[1h])
    )
  )


```

IIUC, the above does the following ( please correct me if am wrong)

1. find the energy consumed in the last 1 hour
```
    (increase(kepler_container_package_joules_total{container_namespace=~\"$namespace\"}[1h]) * $watt_per_second_to_kWh) 
```

calculates the `increase` in the absolute values of `kepler_container_package_joules_total` for the last `1h` and converts that to kWh

2. find the `number of hours` of data that is collected in the last `24h`
```
    (
      count_over_time(kepler_container_package_joules_total{container_namespace=~\"$namespace\"}[24h])
      /
      count_over_time(kepler_container_package_joules_total{container_namespace=~\"$namespace\"}[1h])
    )
```

3. Multiply both (`*`)
4. `sum` all of values to get a single value

Lets we have been measuring for the past 6 hours
1. the cluster has been idle for first 5 hours 
2. in the last 1hr new workloads got deployed which increased the energy consumption quite high (say 20 times the base)

The query above would paint a wrong picture of the usage. The value we get will be way higher than the `sum(increase(kepler_container_package_joules_total[6h]))`

Thus, should be rely on just the `sum( increase( ...) )` than extrapolating the last 1hr 's consumption?

**Comment by marceloamaral** [comment]
Electricity usage is usually shown as the total amount of energy (measured in kilowatt-hours or kWh) used during a specific time, like a day, a month, or a year. The way this information is presented can be customized to suit the needs and preferences of the user or organization. You can get details like how much electricity was used in each hour, the total for a day, average for a month, and more, depending on how detailed or in-depth you want to analyze it.

In this particular query, it's displaying the electricity usage per day, but you can also choose to view it per hour by using just the first part of the function. However, it's important to note that the second part of the function doesn't extrapolate the data to show it for the entire day. Actually, this function was created to fix the issue that you are describing, since the previous functions were just simply multiplying by 24 hours.

The first part of the function counts how many data samples are collected within a 24-hour period, while the second part counts the number of samples collected within a 1-hour period. This is important because, depending on how often Prometheus collects data (scrape interval), the number of samples can vary.

Let's say a workload has 10 data samples within 1 hour, and it runs for 5 hours (as you mentioned in your example). The first part of the function would show 50 samples. When you divide the first part by the second part, you get 5 hours. This means the kWh for this workload would be multiplied by 5 hours, not extrapolated to 24 hours.

**Comment by sthaha** [comment]
@marceloamaral , thanks for chiming in, but my point about the incorrectness still stands .. doesn't it?

> in the last 1hr new workloads got deployed which increased the energy consumption quite high (say 20 times the base)

imagine a graph like this 

<img width="670" alt="image" src="https://github.com/sustainable-computing-io/kepler/assets/3005132/6518a9a2-1f99-48cd-82b2-b5cd99180cff">

Wouldn't the query produce `6 * (40 - 2) kWh` ?

**Reply by marceloamaral** [reply]
Right, the current query is only considering the first 1h interval to extrapolate the results, it should be the rate of 24h.

But the `increase` function actually does that, takes the average value from the range and multiply by the available sample interval.

```
sum by (container_name, pod_name) (
  increase(
   (kepler_container_package_joules_total{}[24h:1m])
  )
) * 0.0000002778
```
where 1J = 1/3600000 = 0.0000002778kWh

Note that, I am defining less steps in the range interval to make the query faster and less heavy to prometheus.

- [24h} Load time: 15298ms  (with 6h time-series)
- [24h:30s} Load time: 10467ms (with 6h time-series)
- [24h:1m} Load time: 8077ms (with 6h time-series)

**Reply by rootfs** [reply]
@williamcaban can you share your thought of how this was designed? thanks

**Comment by sthaha** [comment]
Closing the discussion based on the correction that got merged. 
Thanks a lot for the fix @marceloamaral 🙇
